### ISSUE TITLE
Remove the dependency on Zenhub

### ISSUE THREAD

**Issue by daira** [zcash/developers#41]
Proposal: allow dependencies to be expressed in the initial comment of a ticket, using the following syntax (case-insensitive).

> Blocked-by: <list of [[org/]repo]#number>
> Blocks: <list of [[org/]repo]#number>

Note that unlike Zenhub, this doesn't prevent creating cycles in the graph. What I suggest is to render a general directed graph rather than a DAG, but colour the arrows that are part of a cycle in red. This makes it clear that cycles are an error and makes them easy to spot in order to fix them manually.

To detect cycles in a directed graph, use [Tarjan's strongly connected components algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm), which runs in time $O(|V| + |E|)$.

**Comment by daira** [comment]
This is essential to fully expose the DAG publically and remove one of ECC's information asymmetries.

**Comment by shielded-nate** [comment]
I'm keen on satisfying this ticket (as far as the title) and I want to hash out the implementation more. I propose the following variation to @daira's proposal above:

1. The relationships are specified in the main ticket description instead of a comment. (For this to work well, developers may need to be able to edit ticket descriptions for tickets filed by others.)
2. The format is tweaked to be nicely readable in github flavored markdown.

### Format

The format is a relation header followed by a bulleted md list of ticket references.

This format is chosen for two reasons:

- It's easy to parse.
- It renders nicely for humans.

Relations headers must be an md bolded relation name with a colon on a line by itself, for example `**Blocked-By:**\n`

The relation `Blocked-By` should precede `Blocks` if both are present. (This regularity is to help humans navigate/edit.)

Each list item line must be `- #<NNN>\n` which is a single ticket reference. Github renders these with titles.

#### Markup and Rendering Examples

_Note: the ticket references chosen are just for illustration and have no relation to this ticket._

```
**Blocked-By:**

- #42
- #13
- #1

**Blocks:**

- #2 
- #3
```

-would render as:

**Blocked-By:**

- #42
- #13
- #1

**Blocks:**

- #2 
- #3

### Outstanding Issues

Is allowing both `Blocked-By` and `Blocks` worth it? An ideal tool would always ensure the two sides of the relation are always rendered consistently, but this isn't the ideal tool. So for example ticket `A` could have `Blocked-By: B` and ticket `B` could have no mention of `A` (incomplete but consistent), or `Blocks: A` (the only consistent option), or `Blocked-By: A` (a nonsensical cycle), or both `Blocks: A` and `Blocked-By: A`.

IMO, it may be worth only supporting the `Blocked-By` relation which removes all inconsistency except for cycles (which are mentioned in the original ticket description).

**Comment by shielded-nate** [comment]
I started hunting for pre-existing tools to do this. After all, https://github.com/isaacs/github/issues/210#issuecomment-842660566 was filed in 2014. Here's a quick summary after 10m of poking around:

First, all of the tools I found are only about _tracking_ the dependencies and modifying close/acceptance criteria to require blockers are closed first. None of them generate the nice DAG visualizations of this repo. Given that, here's a quick evaluation of those tools:

- https://github.com/isaacs/github/issues/210#issuecomment-841372091 links to [Actions marketplace: `blocking-issues`](https://github.com/marketplace/actions/blocking-issues) (with latest commit ~1 month ago: https://github.com/Levi-Lesches/blocking-issues/commit/87d4122edf89dc9314d141649b0159a0e4c425d7 ). 👍
- And via https://github.com/isaacs/github/issues/210#issuecomment-842660566 :
  - [dependent-issues action](https://github.com/marketplace/actions/dependent-issues)
    - "looks pretty nice, it's a different approach than @Levi-Lesches by looking into comments for certain keywords. Also supports dependencies between repos. I guess pick the one that suites your use case most."
    - My note: latest commit https://github.com/z0al/dependent-issues/commit/75d554cd9494b6e1766bc9d08a81c26444ad5c5a > 1 year old. 👎 
  - [ZenHub](https://blog.zenhub.com/github-issue-dependencies/) - external tool 👎
  - [CodeTree](https://codetree.com/) - external tool 👎
- Another route is https://github.com/orgs/community/discussions/4928 which mentions the [sub-issue private beta](https://github.com/orgs/community/discussions/131957) - only trees not DAGs 👎

IMO, it's worth considering using the syntax of [Actions marketplace: `blocking-issues`](https://github.com/marketplace/actions/blocking-issues). That looks like it only supports one relation/syntax: `Blocked by: #X, #Y, #Z`. This would allow the two tools to be used together. The features of `blocking-issues` that this tool lacks are:

- Update dependents' tracking comments when dependencies are closed.
- Block PRs from acceptance if any blockers are open.

I guess it does not / cannot block _issues_ from being closed if blockers remain open, which is awkward.

**Comment by Levi-Lesches** [comment]
Seems you accidentally pinged me from that quote, but I'm glad you did! 

> with latest commit ~1 month ago

I don't commit very often, but that's mostly because I think it has reached a "good enough" point where it works without issue on all my repos, and I usually just commit to follow up on complaints or suggestions.

> IMO, it's worth considering using the syntax of [Actions marketplace: blocking-issues](https://github.com/marketplace/actions/blocking-issues). That looks like it only supports one relation/syntax: `Blocked by: #X, #Y, #Z`

Sure, if you want the regex, here it is: 

```js
const regex = /blocked by:? ([#\d, ]+)/ig;  // this is in JS
```

> I guess it does not / cannot block issues from being closed if blockers remain open, which is awkward.

As far as I know, there is no way for a bot to prevent an issue from being closed. Also, even though there is a relatively new "wontfix" close reason, many people still use the standard close to represent wontfix/no response/can't reproduce/already fixed/etc, so blocking issues from closing isn't as important as blocking a PR from being merged when it depends on some other code or features being merged first. 

Let me know if you have any questions, and feel free to take some code from my repo if that helps.

**Comment by nuttycom** [comment]
Now that we've migrated to using the github dependency graph, this can be considered complete.
